include/kodlab_mjbots_sdk/log.h

```diff
@@ -4,199 +4,575 @@
  * For example,
- *
  * ```cpp
  * LOG_WARN("This is a warning message.");
  * LOG_ERROR("%s", "This is an error message.");
  * ```
  *
- * Conditional logging macros are also provided, which take a leading
- * conditional argument before the standard `std::fprintf` input.  For example
- *
+ * Conditional logging macros `LOG_IF_XXXX` are also provided, which take a
+ * leading conditional argument before the standard `std::fprintf` input.  For
+ * example,
  * ```cpp
  * LOG_IF_INFO(false, "%s", "This info message will not be logged.");
  * LOG_IF_FATAL(true, "This fatal message will be logged.");
  * ```
  *
+ * Verbose logging commands are provided for all logging macros, and take the
+ * form `VLOG_XXXX` or `VLOG_IF_XXXX`.  The default output of the log and
+ * verbose log macros is as follows.
+ * ```
+ * [SEVERITY] Log message
+ * [SEVERITY][path/to/file | function:line_no] Verbose log message
+ * ```
+ * The output behavior can be changed by redefining the `LOG_ARGS`,
+ * `LOG_FORMAT`, `VLOG_ARGS`, and `VLOG_FORMAT` macros.  For example, to produce
+ * verbose output of the form
+ * ```
+ * [SEVERITY][path/to/file][line_no][function] Verbose log message
+ * ```
+ * the verbose macros would be redefined as follows
+ * ```cpp
+ * #define VLOG_ARGS(tag) tag, __FILE__, __LINE__, __func__
+ * #define VLOG_FORMAT "[%-6s][%-15s][%d][%s] "
+ * ```
+ *
  * Colored terminal output is provided by default via
  * [ANSI escape codes](https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797).
  * This can be disabled by defining the `NO_COLOR` macro.
  *
+ *
+ * @todo Implement stack trace.
+ * @todo Add time to log output.
+ * @todo Implement logging to file.
+ *
  */
 
 #pragma once
 
-#include <iostream>
+#include <iostream>  // std::fprintf
 
 ///////////////////////////////////////////////////////////////////////////////
 // Console Printing                                                          //
 ///////////////////////////////////////////////////////////////////////////////
 
-// String Formatting
-#define NEWLINE "\n"
+// C Escapes
+#define BELL "\a"  ///< Terminal bell character
+#define BACKSPACE "\b"  ///< Backspace character
+#define TAB "\t"  ///< Horizontal tab character
+#define NEWLINE "\n"  ///< Newline character
+#define RETURN "\r"  ///< Carriage return character
 
 // OStreams
-#define STDERR stderr
-#define STDOUT stdout
-
-// Colors
-#define CONSOLE_SEQ_ESC "\x1b"
-#define CONSOLE_SEQ_BEG "["
-#define CONSOLE_SEQ_SEP ";"
-#define CONSOLE_SEQ_END "m"
-#define FONT_REG "0"
-#define FONT_BOLD "1"
-#define COLOR_RESET "0"
-#define COLOR_BLACK "30"
-#define COLOR_RED "31"
-#define COLOR_GREEN "32"
-#define COLOR_YELLOW "33"
-#define COLOR_BLUE "34"
-#define COLOR_MAGENTA "35"
-#define COLOR_CYAN "36"
-#define COLOR_WHITE "37"
-
-// Color Convenience
+#define STDERR stderr  ///< Standard error stream
+#define STDOUT stdout  ///< Standard output stream
```

Comment:
Some of these more common words may want to have some namespace-like terms added.
`STDERR_LOGH` for example.
Not an issue now, but could be if other files ever define similar macros. 
Leave as is for now.